Implement full Query By Example matching.

bparth24 opened the pull request:
- Creates new queryByExample module with comprehensive field matching logic
- Converts VPR examples to JSON pointers for precise credential filtering
- Updates presentations.js filter chain to use new matching capabilities
- Adds comprehensive unit tests covering all matching scenarios
- Replaces basic context-only filtering with full example-based matching

Previous implementation only matched @context and specific Open Badge fields. New implementation matches ALL fields specified in Query By Example requests using JSON pointer traversal and flexible value comparison logic.

bparth24 committed:
Implement full Query By Example matching.

- Creates new queryByExample module with comprehensive field matching logic
- Converts VPR examples to JSON pointers for precise credential filtering
- Updates presentations.js filter chain to use new matching capabilities
- Adds comprehensive unit tests covering all matching scenarios
- Replaces basic context-only filtering with full example-based matching

Previous implementation only matched @context and specific Open Badge fields.
New implementation matches ALL fields specified in Query By Example requests
using JSON pointer traversal and flexible value comparison logic.

BigBlueHat commented:
@bparth24 can you test this new code with the Academic Course credential's queries both on `main` and the branch at https://github.com/credential-handler/vc-examples/pull/91? My hope is that this patch fixes that issue also.

bparth24 committed:
Remove DEBUG logs

dlongley reviewed (changes requested):
Didn't do a full review yet, but left some comments so things could get kicked off in parallel.

dlongley commented on lib/queryByExample.js:
```diff
@@ -0,0 +1,232 @@
+ * @returns {boolean} True if the value should be used for matching.
+ */
+function _isMatchableValue(value) {
+  // Skip null, undefined
+  if(value == null) {
+    return false;
+  }
```
Hmm, I think we need to be careful here -- perhaps a `null` value means "value must not be present" -- which is different from "ignore this". It's true that `undefined` should be treated as if the associated property wasn't even there to begin with (and generally shouldn't ever happen, but good to cover the case).

dlongley commented on lib/queryByExample.js:
```diff
@@ -0,0 +1,232 @@
+    return false;
+  }
+
+  // Skip empty arrays
+  if(Array.isArray(value) && value.length === 0) {
+    return false;
+  }
```
Hmm, I think empty arrays are wildcards that mean "any array will match". Clearly under-specified in the spec, but I think this was the aim.

dlongley commented on lib/queryByExample.js:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  // Skip empty objects
+  if(typeof value === 'object' && !Array.isArray(value) &&
+      Object.keys(value).length === 0) {
+    return false;
+  }
```
I think empty objects are considered "wildcards" -- meaning, any value will match. Clearly under-specified in the spec, but I think this was the aim.

dlongley commented on lib/queryByExample.js:
```diff
@@ -0,0 +1,232 @@
+  const keys1 = Object.keys(obj1);
+  const keys2 = Object.keys(obj2);
+
+  // Check if they have the same number of keys
+  if(keys1.length !== keys2.length) {
+    return false;
+  }
+
+  // Check if all keys and values match
+  return keys1.every(key =>
+    keys2.includes(key) && _valuesMatch(obj1[key], obj2[key])
+  );
```
I don't think the object matching is "must match this object exactly", but rather, "must include these fields and values". So I suspect this function (or something else) needs an adjustment for this.

In general, if you think about this query mechanism visually, the "example" is to be placed as an overlay on a possible candidate for a match -- and as long as everything in the example "overlaps" with something in the candidate, then the candidate is considered a match.

bparth24 replied:
> I don't think the object matching is "must match this object exactly", but rather, "must include these fields and values". So I suspect this function (or something else) needs an adjustment for this.
> 
> In general, if you think about this query mechanism visually, the "example" is to be placed as an overlay on a possible candidate for a match -- and as long as everything in the example "overlaps" with something in the candidate, then the candidate is considered a match.

@dlongley 

Attempt to address your concern in this commit - https://github.com/digitalbazaar/bedrock-web-wallet/pull/87/commits/c8fad33eda6e04dbd17d9114e8b32ae9048a5265

dlongley commented on lib/queryByExample.js:
```diff
@@ -0,0 +1,232 @@
+    // Check each pointer against the credential content
+    return expectedPointers.every(({pointer, expectedValue}) => {
+      try {
+        const actualValue = JsonPointer.get(credential, pointer);
+        const result = _valuesMatch(actualValue, expectedValue);
```
I think we might need to be more careful here with arrays? We should make sure there are tests with arrays that have objects nested within them that are also arrays still work.

This code might also benefit from using that other code I linked to where just the "deepest" pointers are considered, I'm not sure.

In other words, it might be best to:

1. Generating the deepest pointers for both the example.
2. Get the leaf "expected values" (similar to what the code is already doing).
3. Do some kind of adjustment for "arrays" (pointer values with numbers) -- perhaps multiplex these or save them for special multiphase processing? For example, each pointer that includes an array is really just another "example" that starts at that array and will need to try to match against any element found in the associated array in any candidate match.
4. For each VC, check just these deepest pointers with the appropriate "match" algorithm(s).
5. For each array case, see if there are any "candidate matches" to consider... recursively? Anyway, something to consider.

bparth24 replied:
> I think we might need to be more careful here with arrays? We should make sure there are tests with arrays that have objects nested within them that are also arrays still work.
> 
> This code might also benefit from using that other code I linked to where just the "deepest" pointers are considered, I'm not sure.
> 
> In other words, it might be best to:
> 
> 1. Generating the deepest pointers for both the example.
> 2. Get the leaf "expected values" (similar to what the code is already doing).
> 3. Do some kind of adjustment for "arrays" (pointer values with numbers) -- perhaps multiplex these or save them for special multiphase processing? For example, each pointer that includes an array is really just another "example" that starts at that array and will need to try to match against any element found in the associated array in any candidate match.
> 4. For each VC, check just these deepest pointers with the appropriate "match" algorithm(s).
> 5. For each array case, see if there are any "candidate matches" to consider... recursively? Anyway, something to consider.

@dlongley 

Attempt to address your concern in this commit - https://github.com/digitalbazaar/bedrock-web-wallet/pull/87/commits/c8fad33eda6e04dbd17d9114e8b32ae9048a5265

bparth24 committed:
Optimize Query By Example for batch processing.

Replaces individual credential filtering with efficient batch processing
to address code review feedback on performance. Processes all credentials
at once instead of calling matchCredentials() separately for each item.

- Updates presentations.js to use batch processing approach
- Removes inefficient _matchQueryByExampleFilter function
- Extracts all credential contents, processes in single call
- Maps results back using reference comparison for O(1) lookup
- Adds integration test to verify batch processing functionality
- Removes unused profile setup from integration test

bparth24 commented on lib/presentations.js:
```diff
@@ -418,6 +420,33 @@ function _handleLegacyDraftCryptosuites({presentation}) {
+
+  return ({record: {content}}) => {
+    // Use reusable module to check if this single credential matches
+    const matches = matchCredentials([content], credentialQuery);
```
> It looks like `matchCredentials` should instead just be `matchCredential` and match against a single VC. Callers could then run it over an array if they desired or not -- and clearly this code doesn't need to be using an array (it's just wasted cycles).
> 
> Hmm, but I see below that there's savings in only having to process `credentialQuery` once within the `matchCredentials` call, but we're losing all of that benefit here. So, instead, we need to reorganize the code such that we don't waste this effort. There a few ways to do that that come to mind real quick:
> 
> 1. Create an array of each record's `content` (VCs), run `matchCredentials()` on it, and then run over the records against the results doing an object comparison (shallow or "reference" comparison, not deep) to see which records match. A Map could also be built from content object => record and that could be used in post processing, but I doubt this saves many cycles and is maybe even less efficient that not using one.
> 2. Modifying the `matchCredentials` output to include which indexes matched and do an index match instead of an object comparison.
> 3. Modifying the `matchCredentials` input to accept an object with the `credential` property and return the whole given object in the match results, allowing the object to carry any other arbitrary data for post processing. Post process the results to convert the objects back to records. This seems to leak details about the structure being worked on into the `matchCredentials` API which is suboptimal / probably not a good idea.
> 4. Do something funky with Symbols and Sets or something (probably not unless it's not messy / hard to understand or follow as a maintainer).
> 
> Open to other better ideas too (of course), but idea 1 in the list above (vs. any of the others in that list) keeps the interfaces and separation of concerns the cleanest, I think.

@dlongley 

Attempt to address your concern and going with option 1 in this commit - https://github.com/digitalbazaar/bedrock-web-wallet/pull/87/commits/4d42d8101d3da66afe4dc32794c750f44cca96b9

bparth24 committed:
Add Query By Example (QBE) credential matching with semantic features.

- Implements comprehensive QBE matching system with matchCredentials() function
  supporting semantic match types (mustBeNull, anyArray, anyValue, exactMatch)
- Adds selectiveDisclosureUtil.js with JSON-LD selection and pointer utilities
  including selectJsonLd(), adjustPointers(), and parsePointer() functions
- Integrates array containment matching vs positional matching for flexible
  credential queries supporting overlay matching where credentials can have
  extra fields beyond the example specification
- Handles null value vs missing field distinction with strict semantic rules
- Includes string normalization and type coercion for robust value comparison
- Integrates with existing presentations.match() system while maintaining
  backward compatibility with existing query types
- Adds debug.js utility to conditionally enable/disable console logs during
  development with support for environment variables and browser flags
- Provides comprehensive test suite with 45+ tests covering semantic features,
  edge cases, array matching, nested objects, and real-world scenarios

bparth24 commented on lib/queryByExample.js:
```diff
@@ -0,0 +1,232 @@
+ * @returns {Array} Array of credentials that match the Query By Example
+ *  specification.
+ */
+export function matchCredentials(credentials, queryByExample) {
```
> Use named params in exported APIs:

@dlongley 

Attempt to address your concern in this commit - https://github.com/digitalbazaar/bedrock-web-wallet/pull/87/commits/c8fad33eda6e04dbd17d9114e8b32ae9048a5265

bparth24 commented:
@dlongley cc: @BigBlueHat 

Outputs for newly added tests:

```
presentations.match()
    ✔ should match credentials using Query By Example with batch processing
  queryByExample
    matchCredentials()
      API and Basic Functionality
        ✔ should use named parameters API
        ✔ should return all credentials when no example provided
        ✔ should return all credentials when example is null
        ✔ should handle empty credentials array
      Semantic Features Tests
        ✔ should match credentials with populated arrays
      Empty Object Wildcar (anyValue)
        ✔ should match any value when example has empty object
        ✔ should match populated objects with empty object wildcard
      Null Semantic (mustBeNull)
        ✔ should match only when field is null
        ✔ should match multiple null fields
        ✔ should match individual null fields correctly
        ✔ should match when field is missing
      Overlay Matching
        ✔ should match when credential has extra fields
        ✔ should match nested objects with extra properties
      Array Matching
        ✔ should match single value against array
        ✔ should match array element
        ✔ should match arrays with common elements
        ✔ should match array elements in complex structures
      Complex Nested Structures
        ✔ should handle deep nesting with multiple levels
        ✔ should handle multiple field matching (AND logic)
        ✔ should handle complex nested object matching
      Error Handling and Edge Cases
        ✔ should handle structure mismatch gracefully
        ✔ should handle invalid credentials gracefully
        ✔ should handle complex pointer scenarios
      String Normalization and Type Coercion
        ✔ should handle string trimming
        ✔ should handle string/number coercion
        ✔ should handle reverse number/string coercion
      Real-world Scenarios
        ✔ should handle medical record queries
        ✔ should handle professional license queries
    convertExamplesToPointers()
      Basic Functionality
        ✔ should convert simple example to pointers
        ✔ should include match types for each pointer
      Context Handling
        ✔ should include @context by default
        ✔ should exclude @context when includeContext=false
      Edge Cases
        ✔ should handle empty example
        ✔ should handle null example
        ✔ should handle complex nested structures
      Integration with presentations.match()
        ✔ should work with credential store structure
        ✔ should work with original mockCredential

Finished in 1.151 secs / 0.503 secs @ 18:53:45 GMT-0700 (Pacific Daylight Time)

SUMMARY:
✔ 47 tests completed
All tests passed.
```

dlongley reviewed (commented):
Only a partial review again -- making great progress here!
